dbt/include/snowflake/macros/materializations/table.sql

```diff
@@ -38,7 +38,7 @@
 
 {% endmaterialization %}
 
-{% macro py_write_table(compiled_code, target_relation, temporary=False) %}
+{% macro py_write_table(compiled_code, target_relation, temporary=False, table_type='transient') %}
```

Comment:
I think I'd rather pass `transient` as a boolean into this macro and then handle the translation here for the call to `df.write.mode()`; the translation is specific to this call and we assume this is a boolean in the user config. Right now we have the translation in the `create` macro, which maps to `"transient"` or `""`. Also, is empty string correct here? What happens if we submit a call to `df.write.mode()` with `table_type=""` (versus omitting the kwarg `table_type` entirely)?

Reply:
> Also, is empty string correct here? What happens if we submit a call to `df.write.mode()` with `table_type=""` (versus omitting the kwarg `table_type` entirely)?

@mikealfare that's a wise question 🧠

I didn't test this out personally, but the comments in the snowpark source code describe that case [here](https://github.com/snowflakedb/snowpark-python/blob/100071a27051aa15b1dad98900947e1a5119bbb9/src/snowflake/snowpark/dataframe_writer.py#L146-L148).

So the way it is implemented in this PR currently, the string value of `table_type` just becomes a pass-through in line 55 below:

```python
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', create_temp_table={{temporary}}, table_type='{{table_type}}')
```

Reply:
From [Snowflake docs](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/latest/api/snowflake.snowpark.DataFrameWriter.save_as_table) (and what Doug linked to):

> table_type – The table type of table to be created. The supported values are: temp, temporary, and transient. An empty string means to create a permanent table. Learn more about table types [here](https://docs.snowflake.com/en/user-guide/tables-temp-transient.html).

So effectively:

```py
# Table is transient.
save_as_table(..., table_type='transient')

# Table is non-transient.
save_as_table(..., table_type='')

# Table is non-transient (so default of `table_type` seems to be empty string).
save_as_table(...)
```

Could swap it out to (according to Mike's suggestion):

```sql
-- dbt/include/snowflake/macros/materializations/table.sql
{% macro py_write_table(compiled_code, target_relation, temporary=False, transient=False) %}
...
{% set table_type = 'transient' if transient else '' %}
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', create_temp_table={{temporary}}, table_type='{{table_type}}')
```

Not too sure if it's worth writing additional logic to omit (or not) the `table_type` arg when it is an empty string (or not) - i.e.

```sql
-- dbt/include/snowflake/macros/materializations/table.sql
{% macro py_write_table(compiled_code, target_relation, temporary=False, transient=False) %}
...
{% if transient %}
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', create_temp_table={{temporary}}, table_type='transient')
{% else %}
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', create_temp_table={{temporary}})
```

Happy to follow a suggested pattern @mikealfare

Reply:
Ftr, the docs linked above point out that the kwarg `create_temp_table` is deprecated:

> create_temp_table – (Deprecated) The to-be-created table will be temporary if this is set to True.

Which of course we still use in our `save_as_table()` method calls - so potentially would be a good time to replace that with `table_type` kwarg entirely.

Go from (status quo):
```python
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', create_temp_table={{temporary}})
```

Or (this PR's original intended change):

```python
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', create_temp_table={{temporary}}, table_type='{{table_type}}')
```

To just:
```python
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', table_type='{{table_type}}')
```

And then - we'll have all types via:

```python
# Table is transient.
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', table_type='transient')

# Table is non-transient.
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', table_type='')

# Table is temporary.
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', table_type='temporary')
```

Reply:
Sorry for the delay, just getting around to plowing through my GH notifications. All of this context helps a lot.

Given that SF is deprecating `create_temp_table` in favor of `table_type`, I agree with using `table_type` in the signature, as a string. And I would default it to empty string in alignment with the docs. This also keeps config parsing out of a macro that otherwise does not care about the jinja global `config` that's floating around.

I wouldn't worry about removing the `table_type` argument if `table_type` is not in the call. However, I think we need to be cognizant of backwards compatibility for the `temporary` argument in the `py_write_table` macro. So that means we need to take both arguments in the `py_write_table` macro and combine them. @dbeatty10, correct me if I'm wrong here, but I think that amounts to something like this:

```
{% if temporary == True %}
{% set table_type = "temporary" %}  -- hence override the value of `table_type` that was passed in
{% else %}
-- this else is not needed, but communicates the concept
-- keep the value of `table_type` that was passed in (which could be the default empty string)
{% endif %}
```
Then we update the call to `save_as_table` to exclude `create_temp_table` in alignment with your third option above:
```python
df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', table_type='{{table_type}}')
```

Reply:
That sounds about right @mikealfare.

### Logic

To support both backwards-compatibility as well as forward-facing use cases, I'd suggest that the new `table_type` parameter takes precedence over `temporary` (but only when it is specified).

So like this (in alignment with the Snowflake docs [here](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/latest/api/snowflake.snowpark.DataFrameWriter.save_as_table#:~:text=create_temp_table%20%E2%80%93%20(Deprecated)%20The,types%20here.)):
- use `table_type` when specified
- use `"temporary"` as the table type only when `temporary` is `True` and `table_type` is not specified
- default the table type to `"transient"` if all else fails

e.g., something similar to what you wrote, but with the precedence flipped:

### An untested implementation

```sql
{% macro py_write_table(compiled_code, target_relation, temporary=False, table_type=none) %}

...

{% if table_type is none and temporary %}
    {% set table_type = "temporary" %}
{% elif table_type is none %}
    {# Default to "transient", just like dbt SQL tables in Snowflake #}
    {% set table_type = "transient" %}
{% elif table_type == "permanent" %}
    {# Snowflake treats "" as meaning "permanent" #}
    {% set table_type = "" %}
{% endif %}

...

df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', table_type='{{table_type}}')
```

### Unnecessarily complicated?

This code might appear unnecessarily complicated on first blush. But it's a four-fold consequence of:

1. Goal of giving precedence to `table_type` when a consumer uses the 3-parameter version of the `py_write_table` macro.
1. Goal of being backwards-compatible when a consumer uses the 2-parameter version of the `py_write_table` macro.
1. Goal of using `"transient"` as the default value (unless overridden)
1. Snowflake docs say: "The supported values of `table_type` are: `temp`, `temporary`, and `transient`. An empty string means to create a permanent table."

Reply:
I'm fine with this, but I have one suggestion.

Since we're mimicking the behavior of `py_write_table` and we want folks to use `table_type` moving forward, we want to support `table_type="transient"`, `table_type="temp"`, and `table_type="temporary"`. This logic does actually do that, but I read it multiple times before I realized that happens because we're updating the parameter that comes in, hence it passes straight through. For the sake of readability, especially because it's jinja, I'd like to add an else clause to the if block that just contains the comment "otherwise leave table_type as it is" (or something along those lines). It would save folks some time in the future.

> {# Snowflake treats "" as meaning "permanent" #}
I completely misread the docs. I conflated this with the default value of "transient". It's kind of wild that empty string is a valid value that means something and is also not the default.

Reply:
Wild indeed! 🤠

Your suggestion works for me 👍

Here's what it would look like after adding that piece:

```python
{% macro py_write_table(compiled_code, target_relation, temporary=False, table_type=none) %}

...

{% if table_type is none and temporary %}
    {% set table_type = "temporary" %}
{% elif table_type is none %}
    {#- Default to "transient", just like dbt SQL tables in Snowflake -#}
    {% set table_type = "transient" %}
{% elif table_type == "permanent" %}
    {#- Snowflake treats "" as meaning "permanent" -#}
    {% set table_type = "" %}
{% else %}
    {#- Otherwise leave table_type as it is -#}
{% endif %}

...

df.write.mode("overwrite").save_as_table('{{ target_relation_name }}', table_type='{{table_type}}')
```